src/primitives/sinsemilla.rs

```diff
@@ -0,0 +1,97 @@
+//! The Sinsemilla hash function and commitment scheme.
+
+use std::iter;
+
+use group::{Curve, Group};
+use halo2::{
+    arithmetic::{CurveAffine, CurveExt},
+    pasta::pallas,
+};
+
+const GROUP_HASH_Q: &str = "z.cash:SinsemillaQ";
+const GROUP_HASH_S: &str = "z.cash:SinsemillaS";
+
+const K: usize = 10;
+const C: usize = 253;
+
+fn lebs2ip_32(bits: &[bool]) -> u32 {
+    bits.iter()
+        .enumerate()
+        .fold(0u32, |acc, (i, b)| acc + if *b { 1 << i } else { 0 })
+}
+
+/// Hash extractor for Pallas, from [§ 5.4.8.7].
+///
+/// [§ 5.4.8.7]: https://zips.z.cash/protocol/orchard.pdf#concreteextractorpallas
+fn extract(point: &pallas::Point) -> pallas::Base {
+    // TODO: Should we return the actual bits in a Vec, or allow the caller to use
+    // PrimeField::to_le_bits on the returned pallas::Base?
+    if let Some((x, _)) = point.to_affine().get_xy().into() {
+        x
+    } else {
+        pallas::Base::zero()
+    }
+}
+
+#[allow(non_snake_case)]
+fn Q(domain_prefix: &str) -> pallas::Point {
+    pallas::Point::hash_to_curve(GROUP_HASH_Q)(domain_prefix.as_bytes())
+}
+
+/// `SinsemillaHashToPoint` from [§ 5.4.1.9].
+///
+/// [§ 5.4.1.9]: https://zips.z.cash/protocol/orchard.pdf#concretesinsemillahash
+#[allow(non_snake_case)]
+pub(crate) fn hash_to_point(
+    domain_prefix: &str,
+    msg: impl Iterator<Item = bool> + ExactSizeIterator,
+) -> pallas::Point {
```

Comment:
should this be `pallas::Affine` instead?

Reply:
Agreed, the spec currently defines `P` as the group of points `(x,y)` on the short Weierstrass form of the curve.

![image](https://user-images.githubusercontent.com/552961/109362728-9f010a00-7859-11eb-9d76-a7f4c0bb44f0.png)
![image](https://user-images.githubusercontent.com/552961/109362745-a7f1db80-7859-11eb-8b5e-2e1853421dda.png)

Reply:
`pallas::Point` and `pallas::Affine` are equivalent group elements by construction and have 1:1 conversion (enforced by the `group::Curve` and `group::prime::{PrimeCurve, PrimeCurveAffine}` traits). However, `pallas::Point: group::Curve` is the format that is efficient to perform operations on, and is intended to be what is passed around in-memory. `pallas::Affine` is only intended for encoding/decoding and accessing affine coordinates. Returning `pallas::Affine` from this API would force an unnecessary `projective -> affine -> projective` round-trip on every call to `sinsemilla::commit` or `sinsemilla::short_commit`.

Reply:
I basically retract my comment, as I did basically the same thing in https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/sapling/commitment/pedersen_hashes.rs, even though the [Jubjub group element definitions in the spec](https://zips.z.cash/protocol/protocol.pdf#concretepedersenhash) are also 'just' `(u, v)` etc. 🙈

I would love if the type was named `pallas::ProjectivePoint` or something. 😅

Reply:
What @str4d said. Basically you're always allowed to use isomorphic representations.